### PR TITLE
ppsspp: do not use system ffmpeg as it breaks games

### LINKED ISSUE #84737: PPSSPP should use its own bundled ffmpeg
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
Unrelated
```

### 

- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

I was trying to play games (namely Valkyria Chronicles II) with PPSSPP and used a brew-provided version of PPSSPP.

### What happened (include all command output)?

All the videos either stuttered or were unplayable: showed black screen soon after the first frame. There are multiple error messages in the logs like the following:

```
59:14:295 Odin_Main    E[ME]: HLE/sceMpeg.cpp:1384 UNIMPL sceMpegAvcDecodeFlush(08d571f0)
00:23:691 Odin_Main    E[SCEKERNEL]: HLE/sceKernelThread.cpp:2352 800201a2=sceKernelTerminateThread(2107): already stopped
00:30:271 Odin_Main    E[SCEKERNEL]: HLE/sceKernelThread.cpp:2352 800201a2=sceKernelTerminateThread(2109): already stopped
00:49:131 Odin_Main    E[SCEKERNEL]: HLE/sceKernelThread.cpp:2352 800201a2=sceKernelTerminateThread(2111): already stopped
00:52:142 Odin_Main    E[ME]: HLE/sceMpeg.cpp:1384 UNIMPL sceMpegAvcDecodeFlush(08d571f0)
00:52:243 Odin_Main    E[ME]: HLE/sceMpeg.cpp:1384 UNIMPL sceMpegAvcDecodeFlush(08d571f0)
01:26:443 Odin_Main    E[SCEKERNEL]: HLE/sceKernelThread.cpp:2352 800201a2=sceKernelTerminateThread(2144): already stopped
01:31:434 Odin_Main    E[SCEKERNEL]: HLE/sceKernelThread.cpp:2352 800201a2=sceKernelTerminateThread(2146): already stopped
01:48:812 Odin_Main    E[SCEKERNEL]: HLE/sceKernelThread.cpp:2352 800201a2=sceKernelTerminateThread(2148): already stopped
01:51:797 Odin_Main    E[ME]: HLE/sceMpeg.cpp:1384 UNIMPL sceMpegAvcDecodeFlush(08d571f0)
01:51:896 Odin_Main    E[ME]: HLE/sceMpeg.cpp:1384 UNIMPL sceMpegAvcDecodeFlush(08d571f0)
02:04:811 Odin_Main    E[SCEKERNEL]: HLE/sceKernelThread.cpp:2352 800201a2=sceKernelTerminateThread(2181): already stopped
```

### What did you expect to happen?

I expected the videos to play normally, because I expected brew to use the ffmpeg package bundled within PPSSPP repository. PPSSPP includes a rather [heavily modified ffmpeg version](https://github.com/hrydgard/ppsspp-ffmpeg), so you cannot just use the system/upstream one. I recompiled PPSSPP manually with its own ffmpeg and this worked perfectly. Fortunately their ffmpeg is static.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
1. Open Valkyria Chronicles II in PPSSPP.
2. Let the intro video play.
3. It will stutter quite a lot.
```

### PR DESCRIPTION
**Pull request by vit9696**
closes #84737

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting? — No, see below.
- Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`? — No, see below.

Note: no tests existed in this formula, not adding them either. CI failure is also unrelated to this change.

-----

### CODE REVIEW

**Comment by alebcay** [comment]
Is the patch on lines 14-18 still needed if we're not using system `ffmpeg`?

There's also a failure building on Mojave, looks like `python@3` is needed as a dependency, at least a build time (unsure about runtime). It may or may not be related to this change, but either way, the build is failing due to that so this change cannot be merged as-is.

Regarding the audit failure on Apple Silicon Big Sur, is there an option in `ppsspp` to use system `molten-vk` (or is `ppsspp`'s copy of it heavily patched)? Homebrew already ships `molten-vk`, and it is built appropriately for the target architecture (`x86_64` only on Intel and `arm64` only on Apple Silicon) so we would like to use that if possible. I would say this is not a must-fix in this PR but is something we'd like to see addressed sooner or later (not necessarily by you).

**Comment by vit9696** [comment]
I am unaware of the system MoltenVK option. However, there is a patch to use a newer variant of MoltenVK: https://github.com/hrydgard/ppsspp/pull/13771. Perhaps it is acceptable to use it instead. I should note that I have not tested it though, and in fact I cannot since I do not own Apple Silicon Macs at the moment.

**Comment by vit9696** [comment]
@alebcay is there a good way to perform a dylib update from that PR in brew? I do not think that the ppsspp variant of MoltenVK is any different from upstream. However, a thing to worry about is ABI compatibility. If we just replace the dylib with the one from molten-vk brew port I do not believe it will work reliably as the formula for it gets updated and ppsspp remains not rebuilt.

**Comment by carlocab** [comment]
> I do not think that the ppsspp variant of MoltenVK is any different from upstream.

Let's replace the one in the source tree with a symlink to the one that comes with the `molten-vk` formula then. We may need to do the same with the final install location as well.

> However, a thing to worry about is ABI compatibility. If we just replace the dylib with the one from molten-vk brew port I do not believe it will work reliably as the formula for it gets updated and ppsspp remains not rebuilt.

A test (even a trivial one) would help minimise issues like this.

**Comment by vit9696** [comment]
@carlocab I believe this will likely break, as MoltenVK does not guarantee ABI compatibility. Adding a test here would not really help that as game testing will likely be needed.

**Comment by carlocab** [comment]
> I believe this will likely break, as MoltenVK does not guarantee ABI compatibility.

That's why we need a test.

> Adding a test here would not really help that as game testing will likely be needed.

I think you underestimate how much work the linker does when it loads an application and the libraries it depends on.

**Comment by vit9696** [comment]
@carlocab the test will _not_ help as PPSSPP does not link to libMoltenVK and instead loads it with dlopen only at the time it starts a game as shown [here](https://github.com/hrydgard/ppsspp/blob/0c34c939ad0c8d543c2f791ad348410df32a9a4c/Common/GPU/Vulkan/VulkanLoader.cpp#L283-L290). PPSSPP will run just fine if I entirely remove libMoltenVK.dylib, which I am going to do for arm64 for now.

File: Formula/ppsspp.rb [hunk 1]
```diff
@@ -50,10 +41,17 @@ def install
         prefix.install "PPSSPPSDL.app"
         bin.write_exec_script "#{prefix}/PPSSPPSDL.app/Contents/MacOS/PPSSPPSDL"
         mv "#{bin}/PPSSPPSDL", "#{bin}/ppsspp"
+        # FIXME: Currently bundled libMoltenVK.dylib is not compatible
+        # with Apple Silicon. Remove this when upstreamed.
+        Hardware::CPU.arm? && (rm "#{prefix}/PPSSPPSDL.app/Contents/Frameworks/libMoltenVK.dylib")
```
Comment (carlocab) [inline Formula/ppsspp.rb:46]:
```suggestion
        (prefix/"PPSSPPSDL.app/Contents/Frameworks/libMoltenVK.dylib").unlink
```
We should just remove this unconditionally, then, as we don't want to ship things that we didn't build from source. See https://docs.brew.sh/Acceptable-Formulae#we-dont-like-binary-formulae.

Comment (vit9696) [reply Formula/ppsspp.rb:46]: PPSSPP will not build otherwise.

Comment (carlocab) [reply Formula/ppsspp.rb:46]: I'm confused. You just said this above:

>PPSSPP will run just fine if I entirely remove libMoltenVK.dylib

**Comment by carlocab** [comment]
These two things also seem inconsistent to me:

> I do not believe it will work reliably as the formula for it gets updated and ppsspp remains not rebuilt.

> PPSSPP does not link to libMoltenVK and instead loads it with dlopen only at the time it starts a game

If libMoltenVK is not link and instead only `dlopen`ed, how is rebuilding relevant?

**Comment by vit9696** [comment]
Building PPSSPP without builtin bundled libraries (e.g. ffmpeg) will not work, so it is not about MoltenVK, which is optional, but about the concept. PPSSPP binaries come from a clear source, but building them is non-trivial and impractical.

**Review by carlocab — CHANGES_REQUESTED** [review]
> building them is non-trivial and impractical.

I believe this makes PPSSPP unsuitable for packaging in Homebrew/core, then. See https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-heavy-manual-prepost-install-intervention.

In my view, one of the following needs to be done for this PR to merge:

1. Unconditionally remove the pre-built `libMoltenVK`.
2. Replace the bundled `libMoltenVK` with the one from the `molten-vk` formula.
3. Build the necessary version of `libMoltenVK` from source in this formula.
4. Add a `disable!` line to the formula, potentially for a future date to account for a deprecation period, because PPSSPP is not suitable for Homebrew/core.

**Comment by vit9696** [comment]
Sorry, but PPSSPP will continue to bundle .a and .dylib files just because we have a discussion here, likely because brew is not the only thing to build it. The problem is not really (just) with libMoltenVK. If you do not plan to disregard this at least for the scope of this PR as it does *not* change the state of the art, I am fine to abandon this change.

**Comment by carlocab** [comment]
> Sorry, but PPSSPP will not stop to bundle .a and .dylib files just because we have a discussion here, likely because brew is not the only thing to build it.

That's fine. Where did I say PPSSPP needs to stop bundling these, or that PPSSPP needs to make changes to their build system in any way? All changes I'm proposing are for the formula hosted in this repository. I am not suggesting any changes be made outside this repo.

> The problem is not really (just) with libMoltenVK.

What else is the problem then?

> it does _not_ change the state of the art

I have no idea what this means.

**Comment by vit9696** [comment]
Alright, done.

**Comment by Gcenx** [comment]
It really would be a good idea to try making use of brew molten-vk formula, the closed PR for updating MoltenVK just replaced the prebuilt binaries from the Vulkan SDK and tweaked the directory structure to match.

I’ve not seen any breakages from bumping MoltenVK yet and I’ve been doing this constantly from MoltenVK v1.1.0 for DXVK.

**Comment by vit9696** [comment]
While I agree that it may be nice to add Vulkan to PPSSPP homebrew formula, I do not think it relates to this PR. Having Vulkan is neither essential nor beneficial in many cases, as the games may run less stable in the first place.

**Comment by Gcenx** [comment]
Ad the changing affects all users who have previously been using the Vulkan backend it would be good to at least add a note about this change, or only make this change for Apple Silicon users.

**Comment by vit9696** [comment]
This is exactly what I did earlier, but apparently there is a strong fight on that from @carlocab.

As I mentioned there is no simple way to test PPSSPP to be compatible with libMoltenVK.dylib coming from homebrew. As a result the choice is between:
1. Dropping a half-broken feature that did not work on arm64 earlier and was imperfect for x86_64.
2. Adding the feature to randomly break in the future or, even worse, causing game issues that are not easy to trace.

My personal opinion is that we should opt with 1 and eventually remove a hack with dylib removal once arm64 MoltenVK support lands to PPSSPP.

**Comment by Gcenx** [comment]
Maybe I didn’t explain this well?

I’m saying only unlink MoltenVK on Apple Silicon, let’s not bother messing with using the brew molten-vk formula for now.

**Comment by vit9696** [comment]
@Gcenx, you did explain that well, but you should address your question not to me but @carlocab. I did exactly what you suggested in my [initial commit](https://github.com/Homebrew/homebrew-core/pull/84738/commits/c1a8fdc78793aba451e1ee37dd030131894fc95c). However, @carlocab, who I believe speaks by the name of @Homebrew here, disagrees with this route. For this reason I [committed](https://github.com/Homebrew/homebrew-core/pull/84738/commits/9ca5e282f72b4beb3f06df736ecdc357b5a8596d) exactly what he suggested [here](https://github.com/Homebrew/homebrew-core/pull/84738#discussion_r703346952).

I do not have an intention to explain what is the right way to do here, especially since I personally build PPSSPP from source and do not own Apple Silicion devices. My only point is to induce less harm to the end users compared to what we already have.

**Comment by Gcenx** [comment]
I didn’t read the conversation under resolved this makes sense now, it’s not allowed due to how brew expects things to function homebrew-core.

I’ve used upstream MoltenVK on projects that were built with much older versions of MoltenVK and not seen any breakages a test shouldn’t be required.

I’d assume the bundled ffmpeg would also fall into this, it would also need to be compiled from source.

**Comment by vit9696** [comment]
My personal opinion is that ppsspp should be cask, because I find it extremely weird that a bundle app is installed into some weird place and is not added to `/Applications`. However, this would require some basic interaction with upstream.

**Comment by unknownbrackets** [comment]
> PPSSPP binaries come from a clear source, but building them is non-trivial and impractical.

I don't want to necessarily get in the way of updates or merges on this pull.  But, ignoring MoltenVK for a second, I agree it would be best to build the included FFmpeg version from source.

Originally, ppsspp-ffmpeg existed only to include prebuilt Windows and Android binaries, mainly because we wanted to keep the build instructions very obvious and easy.  FFmpeg's build steps on Windows are anything but.

Later, binaries for other mobile systems (including Blackberry, etc.) were included.  Eventually, this also led to more binaries for Linux/macOS/etc., under a probably misguided sense of consistency and drive to reduce CI build times.

There's no very good reason to use those specific binaries.  There *are* very good reasons to use that specific FFmpeg version (see #84737.)

-[Unknown]

**Comment by alebcay** [comment]
I've updated the pull request to build the bundled `ffmpeg` rather than use prebuilt artifacts in the bundled `ffmpeg` folder. From test results looks like I might have to take a look at the build script for bundled `ffmpeg` on arm64. Maybe we just need to build a fat intermediate despite the fact that we're not building a fat target...

Also removed `glew` as a dependency since [upstream explicitly ignores system `glew` on macOS](https://github.com/hrydgard/ppsspp/commit/63f81f96fdf1dfa4f8bee76a809b3864926dd669).

Removed `libMoltenVK` and `libsdl2` dylibs from inside the app bundle and replaced them with symlinks to Homebrew-managed library paths instead. Verified that running `ppsspp` with the symlink to Homebrew `libMoltenVK` seems to work (the `dlopen` succeeds):
```
VulkanMayBeAvailable: Library loaded ('@executable_path/../Frameworks/libMoltenVK.dylib')
[...]
VulkanMayBeAvailable: Calling vkCreateInstance
[mvk-info] MoltenVK version 1.1.5, supporting Vulkan version 1.1.189.
[...]
VulkanMayBeAvailable: Vulkan test instance created successfully.
VulkanMayBeAvailable: Found working Vulkan API!
VulkanMayBeAvailable: Destroying instance
DEBUG: Vulkan might be available.
```

Fixed a few flags to enforce usage of system `snappy`/`libzip`. Added `zlib` as Linux dependency (system `zlib` will be used automatically if present).

Upstream also ships [a number of other external/vendored dependencies](https://github.com/hrydgard/ppsspp/tree/master/ext), some of which cannot be toggled to use system versions. Specifically w.r.t. Homebrew, looks like they ship or add as modules specific versions of `glew`, `vulkan-headers`, `cityhash`, `zstd`, `spirv-cross`.

I think this is about as good as we can get right now with using Homebrew-provided dependencies where possible, short of stripping out upstream's entire build system like was done for [the `ccextractor` formula](https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/ccextractor.rb) via [a big patch](https://raw.githubusercontent.com/Homebrew/formula-patches/9bc4ef5a88b9a4d55dead30130aa79f8eee5faf7/ccextractor/unbundle-libs.patch).

To echo [@carlocab's thoughts from above](https://github.com/Homebrew/homebrew-core/pull/84738#pullrequestreview-747776326), possibly of concern here is still Homebrew's policy of not packaging [stuff that requires vendored versions of Homebrew formulae](https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae). However, [there's always room for exceptions](https://docs.brew.sh/Acceptable-Formulae#sometimes-there-are-exceptions) if we feel it's appropriate.

> My personal opinion is that ppsspp should be cask, because I find it extremely weird that a bundle app is installed into some weird place and is not added to /Applications. However, this would require some basic interaction with upstream.

I think this would be the ideal outcome. It would allow for PPSSPP to be accessible within the Homebrew ecosystem but still allow for it to be built and distributed on upstream's terms. However, this requires upstream to provide binaries/builds for macOS, which [they currently don't](https://ppsspp.org/downloads.html).

**Comment by Gcenx** [comment]
Need to make this require Xcode11.7 minimum or macOS Catalina and later only due to `molten-vk` having those requirements.

**Comment by alebcay** [comment]
The Xcode 11.7 requirement is for building Molten-VK and is [already expressed in the `molten-vk` formula](https://github.com/Homebrew/homebrew-core/blob/96b79c0e3776caa530fc2b3fde57f5c8b5479f0c/Formula/molten-vk.rb#L16).

Does Molten-Vk depend on Catalina or later? Right now [the formula indicates that we only need Sierra or later](https://github.com/Homebrew/homebrew-core/blob/96b79c0e3776caa530fc2b3fde57f5c8b5479f0c/Formula/molten-vk.rb#L18) and [Molten-VK's Runtime User Guide](https://github.com/KhronosGroup/MoltenVK/blob/master/Docs/MoltenVK_Runtime_UserGuide.md) also indicates that "at runtime, MoltenVK requires at least macOS 10.11, iOS 9, or tvOS 9 (or iOS 11 or tvOS 11 if using IOSurfaces)".

**Comment by Gcenx** [comment]
The Xcode11.7 requirement is for compiling it can run on 10.11 just won’t have as many working features as 10.13/10.14 system.

You can get Xcode11.7 working on macOS Mojave using retroactive didn’t test below this.

File: Formula/ppsspp.rb [hunk 2]
```diff
@@ -28,31 +17,64 @@ class Ppsspp < Formula
...
       if OS.mac?
         prefix.install "PPSSPPSDL.app"
         bin.write_exec_script "#{prefix}/PPSSPPSDL.app/Contents/MacOS/PPSSPPSDL"
         mv "#{bin}/PPSSPPSDL", "#{bin}/ppsspp"
+
+        # Replace app bundles with symlinks to allow dependencies to be updated
+        (prefix/"PPSSPPSDL.app/Contents/Frameworks/libMoltenVK.dylib").unlink
+        (prefix/"PPSSPPSDL.app/Contents/Frameworks/libSDL2-2.0.0.dylib").unlink
+        (prefix/"PPSSPPSDL.app/Contents/Frameworks").install_symlink Formula["molten-vk"].opt_lib/"libMoltenVK.dylib"
+        (prefix/"PPSSPPSDL.app/Contents/Frameworks").install_symlink Formula["sdl2"].opt_lib/"libSDL2-2.0.0.dylib"
```
Comment (carlocab) [inline Formula/ppsspp.rb:71]:
```suggestion
        app_frameworks = prefix/"PPSSPPSDL.app/Contents/Frameworks"
        ln_sf (Formula["molten-vk"].opt_lib/"libMoltenVK.dylib").relative_path_from(app_frameworks), app_frameworks
        ln_sf (Formula["sdl2"].opt_lib/"libSDL2-2.0.0.dylib").relative_path_from(app_frameworks), app_frameworks
```
`install_symlink` will create one that references Cellar paths, and this gets broken on `molten-vk`/`sdl2` version/revision bumps. We don't want that.

File: Formula/ppsspp.rb [hunk 3]
```diff
@@ -28,31 +17,64 @@ class Ppsspp < Formula
...
       else
         bin.install "PPSSPPSDL" => "ppsspp"
       end
     end
   end
+
+  test do
+    system "#{bin}/ppsspp", "--version"
```
Comment (carlocab) [inline Formula/ppsspp.rb:79]:
```suggestion
    system "#{bin}/ppsspp", "--version"
    
    if OS.mac?
      app_frameworks = prefix/"PPSSPPSDL.app/Contents/Frameworks"
      assert_predicate app_frameworks/"libMoltenVK.dylib", :exist?, "Broken linkage with `molten-vk`"
      assert_predicate app_frameworks/"libSDL2-2.0.0.dylib", :exist?, "Broken linkage with `sdl2`"
    end
```
Let's check that these symlinks aren't stale. The `sdl2` one probably isn't needed, but I guess it doesn't hurt.

**Comment by alebcay** [comment]
Homebrew really only tests and builds against the three most recent macOS versions (right now, back to Mojave) anyways, so there's no need to really document anything under that.

**Comment by Gcenx** [comment]
> Homebrew really only tests and builds against the three most recent macOS versions (right now, back to Mojave) anyways, so there's no need to really document anything under that.

It’s more I’m less familiar with how the brew testbot functions, I see that’s the 10.14 didn’t get counted as a failure.

File: Formula/ppsspp.rb [hunk 4]
```diff
@@ -28,31 +17,70 @@ class Ppsspp < Formula
...
+        ln_sf (Formula["sdl2"].opt_lib/"libSDL2-2.0.0.dylib").relative_path_from(app_frameworks), app_frameworks
       else
         bin.install "PPSSPPSDL" => "ppsspp"
       end
     end
   end
+
+  test do
+    system "#{bin}/ppsspp", "--version"
+
```
Comment (carlocab) [inline Formula/ppsspp.rb:79]:
```suggestion

```
Ugh, sorry. The GitHub UI sucks.

**Comment by alebcay** [comment]
@carlocab, any thoughts on the Linux build failure? It was working within one of the most recent couple of commits.

**Comment by carlocab** [comment]
Still needs `glew`, maybe?

**Comment by alebcay** [comment]
It should have just used it's bundled (static) `glew` instead on Linux but added it back and hopefully that'll fix it.

**Comment by carlocab** [comment]
LGTM. Should we squash all these commits together so that @BrewTestBot doesn't squash them all into the uninformative `ppsspp: revision`?

**Comment by alebcay** [comment]
Done. Please thoroughly review the changes again; there was a merge commit mixed into the original changes so I had to manually re-pick some commits and resolve some conflicts before squashing. I'd like to think that nothing bad happened but you never know 🙂

I do hope that upstream figures out their macOS CI situation and is able to provide binaries in the future. I still think PPSSPP makes a lot more sense as a cask, given how it handles dependencies and is a GUI app.

**Review by carlocab — APPROVED** [review]

**Pull request closed without merging**